### PR TITLE
spectests: Refactor value reading and checking

### PR DESCRIPTION
**Pull request by chfast**

### CODE REVIEW

File: test/spectests/spectests.cpp [hunk 1]
```diff
@@ -525,7 +522,7 @@ class test_runner
             return false;
         }
 
-        const auto expected_value = read_floating_point_value<T>(expected);
+        const auto expected_value = read_value(expected)->value;  // TODO: Type ignored.
```
Comment (gumb0) [inline test/spectests/spectests.cpp:525]: Is this comment still relevant? (second commits removes the same comment for integers, but the code is quite similar)

File: test/spectests/spectests.cpp [hunk 2]
```diff
@@ -542,10 +539,14 @@ class test_runner
 
     bool check_result(fizzy::Value actual_value, const json& expected)
     {
+        // TODO: Check types here.
+        // TODO: Use read_value() here too if it can handle "nan:..." strings.
```
Comment (gumb0) [inline test/spectests/spectests.cpp:543]: Well `Value` can't represent expectation like "arithmetic NaN expected"

File: test/spectests/spectests.cpp [hunk 3]
```diff
@@ -406,40 +407,33 @@ class test_runner
...
-    std::optional<fizzy::Value> read_value(const json& v)
-    {
-        const auto type = v.at("type").get<std::string>();
+        const uint64_t raw_value = std::stoull(v.at("value").get<std::string>());
 
-        if (type == "i32" || type == "i64")
-            return read_integer_value(v);
+        if (type == "i32")
+        {
+            assert(static_cast<uint32_t>(raw_value) == raw_value && "overflow if i32 test value");
```
Comment (gumb0) [inline test/spectests/spectests.cpp:419]:
```suggestion
            assert(static_cast<uint32_t>(raw_value) == raw_value && "overflow in i32 test value");
```

File: test/spectests/spectests.cpp (hunk 2 above)
Comment (chfast) [reply test/spectests/spectests.cpp:543]: True. But you can extend the Value to handle that or handle `nan:...` first using strings and then `read_value()`.

File: test/spectests/spectests.cpp (hunk 1 above)
Comment (chfast) [reply test/spectests/spectests.cpp:525]: Yes. read_value() produces type and value, but type is discarded here.

File: test/spectests/spectests.cpp [hunk 4]
```diff
@@ -406,40 +407,33 @@ class test_runner
...
+            return static_cast<uint32_t>(raw_value);
+        }
+        else if (type == "i64")
+        {
+            return raw_value;
+        }
         else if (type == "f32")
-            return read_floating_point_value<float>(v);
+        {
+            assert(static_cast<uint32_t>(raw_value) == raw_value && "overflow if f32 test value");
```
Comment (axic) [inline test/spectests/spectests.cpp:428]:
```suggestion
            assert(static_cast<uint32_t>(raw_value) == raw_value && "overflow in f32 test value");
```

File: test/spectests/spectests.cpp (hunk 3 above)
Comment (axic) [reply test/spectests/spectests.cpp:419]: What exactly is the string part doing? It i just a true so no op. Why not leave it as a comment instead?

**Review by axic — APPROVED** [review]
Looks good apart from those smaller comments.

**Review by gumb0 — APPROVED** [review]

File: test/spectests/spectests.cpp (hunk 3 above)
Comment (gumb0) [reply test/spectests/spectests.cpp]: It's a comment, that will be in the output when assertion fails, I think. Could be a regular commit, but I don't mind merging it like this.

Comment (axic) [reply test/spectests/spectests.cpp]: Ah right, it will be "literalized". Yeah, could keep it this way.

**Pull request merged**